File: src/nodetool/models/secret.py
"""
Defines the Secret database model for encrypted secret storage.

Secrets are encrypted using a master key and user_id as salt, providing
per-user encryption isolation.
"""

from datetime import UTC, datetime
from typing import Optional

from nodetool.models.base_model import DBField, DBIndex, DBModel, create_time_ordered_uuid
from nodetool.models.condition_builder import Field


@DBIndex(["user_id", "key"], unique=True)
@DBIndex(["user_id"])
class Secret(DBModel):
    """Database model for encrypted secrets."""

    @classmethod
    def get_table_schema(cls):
        """Returns the database table schema for secrets."""
        return {"table_name": "nodetool_secrets"}

    id: str = DBField(default_factory=create_time_ordered_uuid)
    user_id: str = DBField()
    key: str = DBField()
    encrypted_value: str = DBField()
    description: Optional[str] = DBField(default=None)
    created_at: datetime = DBField(default_factory=datetime.now)
    updated_at: datetime = DBField(default_factory=datetime.now)

    def before_save(self):
        """Updates the `updated_at` timestamp before saving."""
        self.updated_at = datetime.now()

    @classmethod
    async def create(cls, user_id: str, key: str, value: str, description: Optional[str] = None, **kwargs):  # type: ignore[override]
        """
        Create a new encrypted secret.

        Args:
            user_id: The ID of the user who owns this secret.
            key: The key/name for this secret (e.g., "OPENAI_API_KEY").
            value: The plaintext value to encrypt and store.
            description: Optional description of the secret.
            **kwargs: Additional fields to set on the model.

        Returns:
            The newly created Secret instance.
        """
        from nodetool.security.crypto import SecretCrypto
        from nodetool.security.master_key import MasterKeyManager

        # Get master key and encrypt the value
        master_key = await MasterKeyManager.get_master_key()
        encrypted_value = SecretCrypto.encrypt(value, master_key, user_id)

        return await super().create(
            id=create_time_ordered_uuid(),
            user_id=user_id,
            key=key,
            encrypted_value=encrypted_value,
            description=description,
            created_at=datetime.now(),
            updated_at=datetime.now(),
            **kwargs,
        )

    @classmethod
    async def find(cls, user_id: str, key: str) -> Optional["Secret"]:
        """
        Find a secret by user_id and key.

        Args:
            user_id: The user ID.
            key: The secret key.

        Returns:
            The Secret instance if found, None otherwise.
        """
        condition = Field("user_id").equals(user_id).and_(Field("key").equals(key))
        results, _ = await cls.query(condition, limit=1)
        return results[0] if results else None

    @classmethod
    async def list_for_user(
        cls, user_id: str, limit: int = 100, start_key: Optional[str] = None
    ) -> tuple[list["Secret"], str]:
        """
        List all secrets for a user.

        Args:
            user_id: The user ID.
            limit: Maximum number of secrets to return.
            start_key: Pagination key for continuation.

        Returns:
            A tuple of (list of Secret instances, next pagination key).
        """
        condition = Field("user_id").equals(user_id)

        if start_key:
            condition = condition.and_(Field("id").greater_than(start_key))

        return await cls.query(condition, limit=limit)

    @classmethod
    async def list_all(cls, limit: int = 1000) -> list["Secret"]:
        """Return all secrets (best effort, limited by the provided limit)."""
        secrets, _ = await cls.query(limit=limit)
        return secrets

    async def get_decrypted_value(self) -> str:
        """
        Decrypt and return the secret value.

        Returns:
            The decrypted plaintext value.

        Raises:
            Exception: If decryption fails (e.g., wrong master key).
        """
        from nodetool.security.crypto import SecretCrypto
        from nodetool.security.master_key import MasterKeyManager

        master_key = await MasterKeyManager.get_master_key()
        return SecretCrypto.decrypt(self.encrypted_value, master_key, self.user_id)

    async def update_value(self, new_value: str) -> None:
        """
        Update the secret value.

        Args:
            new_value: The new plaintext value to encrypt and store.
        """
        from nodetool.security.crypto import SecretCrypto
        from nodetool.security.master_key import MasterKeyManager

        master_key = await MasterKeyManager.get_master_key()
        self.encrypted_value = SecretCrypto.encrypt(new_value, master_key, self.user_id)

        # Invalidate cache
        from nodetool.security.secret_helper import clear_secret_cache

        clear_secret_cache(self.user_id, self.key)

        await self.save()

    @classmethod
    async def upsert_encrypted(
        cls,
        user_id: str,
        key: str,
        encrypted_value: str,
        description: Optional[str] = None,
        created_at: Optional[datetime] = None,
        updated_at: Optional[datetime] = None,
    ) -> "Secret":
        """
        Create or update a secret using a pre-encrypted value.

        This is used when migrating secrets between instances that share the same
        master key. The encrypted blob is persisted as-is without decrypting.
        """
        existing = await cls.find(user_id, key)
        if created_at is None:
            created_at = datetime.now(UTC)
        if updated_at is None:
            updated_at = datetime.now(UTC)

        if existing:
            existing.encrypted_value = encrypted_value
            existing.updated_at = updated_at
            if description is not None:
                existing.description = description

            # Invalidate cache
            from nodetool.security.secret_helper import clear_secret_cache

            clear_secret_cache(user_id, key)

            await existing.save()
            return existing

        return await super().create(
            id=create_time_ordered_uuid(),
            user_id=user_id,
            key=key,
            encrypted_value=encrypted_value,
            description=description,
            created_at=created_at,
            updated_at=updated_at,
        )

    @classmethod
    async def upsert(cls, user_id: str, key: str, value: str, description: Optional[str] = None) -> "Secret":
        """
        Create or update a secret.

        If a secret with the same user_id and key exists, update it.
        Otherwise, create a new one.

        Args:
            user_id: The user ID.
            key: The secret key.
            value: The plaintext value to encrypt and store.
            description: Optional description.

        Returns:
            The Secret instance (created or updated).
        """

        existing = await cls.find(user_id, key)

        if existing:
            await existing.update_value(value)
            if description is not None:
                existing.description = description
                await existing.save()

            # Invalidate cache
            from nodetool.security.secret_helper import clear_secret_cache

            clear_secret_cache(user_id, key)

            return existing
        else:
            return await cls.create(user_id=user_id, key=key, value=value, description=description)

    @classmethod
    async def delete_secret(cls, user_id: str, key: str) -> bool:
        """
        Delete a secret.

        Args:
            user_id: The user ID.
            key: The secret key.

        Returns:
            True if the secret was deleted, False if not found.
        """
        secret = await cls.find(user_id, key)
        if secret:
            await secret.delete()

            # Invalidate cache
            from nodetool.security.secret_helper import clear_secret_cache

            clear_secret_cache(user_id, key)

            return True
        return False

    def to_dict_safe(self) -> dict:
        """
        Return a dictionary representation without the encrypted value.

        This is safe to return in API responses.

        Returns:
            A dictionary with id, user_id, key, description, created_at, updated_at.
        """
        return {
            "id": self.id,
            "user_id": self.user_id,
            "key": self.key,
            "description": self.description,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }

File: ts/packages/models/src/secret.ts
/**
 * Secret model – stores encrypted secrets per user.
 *
 * Port of Python's `nodetool.models.secret`.
 */

import type { TableSchema, Row } from "./database-adapter.js";
import {
  DBModel,
  createTimeOrderedUuid,
  type IndexSpec,
  type ModelClass,
} from "./base-model.js";
import { field } from "./condition-builder.js";
import { encrypt, decrypt, getMasterKey } from "@nodetool/security";

// ── Schema ───────────────────────────────────────────────────────────

const SECRET_SCHEMA: TableSchema = {
  table_name: "nodetool_secrets",
  primary_key: "id",
  columns: {
    id: { type: "string" },
    user_id: { type: "string" },
    key: { type: "string" },
    encrypted_value: { type: "string" },
    description: { type: "string", optional: true },
    created_at: { type: "string" },
    updated_at: { type: "string" },
  },
};

const SECRET_INDEXES: IndexSpec[] = [
  { name: "idx_secrets_user_key", columns: ["user_id", "key"], unique: true },
  { name: "idx_secrets_user_id", columns: ["user_id"], unique: false },
];

// ── Model ────────────────────────────────────────────────────────────

export class Secret extends DBModel {
  static override schema = SECRET_SCHEMA;
  static override indexes = SECRET_INDEXES;

  declare id: string;
  declare user_id: string;
  declare key: string;
  declare encrypted_value: string;
  declare description: string;
  declare created_at: string;
  declare updated_at: string;

  constructor(data: Row) {
    super(data);
    const now = new Date().toISOString();
    this.id ??= createTimeOrderedUuid();
    this.description ??= "";
    this.created_at ??= now;
    this.updated_at ??= now;
  }

  override beforeSave(): void {
    this.updated_at = new Date().toISOString();
  }

  /**
   * Find a secret by user_id and key.
   */
  static async find(userId: string, key: string): Promise<Secret | null> {
    const condition = field("user_id").equals(userId).and(field("key").equals(key));
    const [results] = await (Secret as unknown as ModelClass<Secret>).query({
      condition,
      limit: 1,
    });
    return results.length > 0 ? results[0] : null;
  }

  /**
   * Create or update a secret.
   *
   * Encrypts the plaintext value before storing.
   */
  static async upsert(opts: {
    userId: string;
    key: string;
    value: string;
    description?: string;
  }): Promise<Secret> {
    const masterKey = getMasterKey();
    const encryptedValue = encrypt(masterKey, opts.userId, opts.value);
    const now = new Date().toISOString();

    const existing = await Secret.find(opts.userId, opts.key);

    if (existing) {
      existing.encrypted_value = encryptedValue;
      existing.updated_at = now;
      if (opts.description !== undefined) {
        existing.description = opts.description ?? "";
      }
      await existing.save();
      return existing;
    }

    return (await (Secret as unknown as ModelClass<Secret>).create({
      id: createTimeOrderedUuid(),
      user_id: opts.userId,
      key: opts.key,
      encrypted_value: encryptedValue,
      description: opts.description ?? "",
      created_at: now,
      updated_at: now,
    })) as Secret;
  }

  /**
   * Delete a secret by user_id and key.
   *
   * @returns True if the secret was deleted, false if not found.
   */
  static async deleteSecret(userId: string, key: string): Promise<boolean> {
    const secret = await Secret.find(userId, key);
    if (secret) {
      await secret.delete();
      return true;
    }
    return false;
  }

  /**
   * List all secrets for a user.
   */
  static async listForUser(
    userId: string,
    limit = 100
  ): Promise<[Secret[], string]> {
    const condition = field("user_id").equals(userId);
    return (Secret as unknown as ModelClass<Secret>).query({
      condition,
      limit,
    });
  }

  /**
   * Get the decrypted plaintext value.
   */
  async getDecryptedValue(): Promise<string> {
    const masterKey = getMasterKey();
    return decrypt(masterKey, this.user_id, this.encrypted_value);
  }

  /**
   * Return a safe dictionary representation without the encrypted value.
   */
  toSafeObject(): Record<string, unknown> {
    return {
      id: this.id,
      user_id: this.user_id,
      key: this.key,
      description: this.description,
      created_at: this.created_at,
      updated_at: this.updated_at,
    };
  }
}
